### ISSUE TITLE
Benchmarking of GeoDataset for a paper result

### ISSUE THREAD

**Issue by calebrob6** [microsoft/torchgeo#81]
### Datasets

We want to test several popular image sources, as well as both raster and vector labels.

- [ ] NAIP + Chesapeake
- [ ] Landsat + CDL
- [ ] Sentinel + Canadian Building Footprints

There is also a question of which file formats to test. For example, sampling from GeoJSON can take 3 min per getitem, whereas ESRI Shapefile only takes 1 sec per getitem (https://github.com/microsoft/torchgeo/pull/69#issuecomment-892816306).

### Experiments

- [ ] **Data location:** local vs. blob storage
- [ ] **Sampling strategy:** random vs. 2-step random tile/chip vs. non-random grid sampling
- [ ] **Warping strategy:** on-the-fly vs. pre-processing step
- [ ] **I/O strategy:** load/warp entire file vs. load/warp/cache entire file vs. load/warp single window

For the warping strategy, we should test the following possibilities:

- [ ] Already in correct CRS/resolution
- [ ] Need to change CRS
- [ ] Need to change resolution
- [ ] Need to change both CRS and resolution

What is the upfront cost of these pre-processing steps?

Example notebook: https://gist.github.com/calebrob6/d9bc5609ff638d601e2c35a1ab0a2dec

**Comment by calebrob6** [comment]
@adamjstewart, sketch of the full experiment:
- Get Landsat scenes from several projections and CDL data.
  - Convert all to COG if they aren't already
- Copy the files to blob storage, local SSD, local HDD
- Create three GeoDataset instances, one for each dataset location
- Record the number of patches per second you can read using each of the GeoDatasets with the different types of GeoSamplers
- Record how long it takes to warp/reproject the Landsat scenes to align them to CDL (or vice-versa) "by hand" with gdalwarp. 
  - Also record the size of the resulting files.
  - Also record the nasty gdalwarp command you actually have to figure out and execute to do this.
  - Note: We can use this to extrapolate how much preprocessing you would need to do before training with a traditional DL library.
  - Note: We can use this pre-aligned data with a custom dataloader to see how many patches/second you could sample if you did go in and do all the preprocessing. Hopefully this number is similar to what you get with torchgeo (or at least not much much larger).

**Comment by adamjstewart** [comment]
@calebrob6 the above proposal covers the matrix of:

* **Data location:** local SSD, local HDD, blob storage
* **Sampling strategy:** RandomGeoSampler, RandomBatchGeoSampler, GridGeoSampler

There are a lot of additional constraints that we're currently skipping:

* **File format:** GeoTIFF vs. HDF5, Shapefile vs. GeoJSON
* **Warping strategy:** already in correct CRS/res, change CRS, change res, change CRS and res

Do you think it's fine to skip these for the sake of time? I doubt reviewers would straight up reject us for not including one of these permutations, and can always ask us to perform additional experiments if they want.

Also, we should definitely benchmark not only `RasterDataset` but also `VectorDataset` (maybe Sentinel + Canadian Building Footprints?). Should I purposefully change the resolution of one of these datasets? Should I purposefully switch to a CRS different than all files or keep the CRS of one of the files?

**Comment by adamjstewart** [comment]
Also, do we want to compare with different batch_sizes or different num_workers?

**Comment by calebrob6** [comment]
I'd do the first matrix as quickly as possible because the results of that are going to be very informative. If that all works out then you can repeat the same with a vectordataset.

> File format: GeoTIFF vs. HDF5, Shapefile vs. GeoJSON
I don't think this is important right now. I.e. we can just assume the data is in a good format (COG and shapefile/geopackage)

> Warping strategy
In the above sketch you can repeat the experiments with the manually aligned versions of the dataset to test the "already in correct CRS/res" case. The first set of experiments is with "change CRS and res". It might be interesting to see if warping or resampling is more expensive, but not interesting for the paper I think.

**Comment by calebrob6** [comment]
Some things to discuss soon:
- How to benchmark CDL/Landsat from blob containers
- How to compare patches/sec to something that people might understand (`torchvision.datasets.ImageNet` seems like a good idea)

**Comment by adamjstewart** [comment]
We're following up on this discussion in https://github.com/microsoft/torchgeo/issues/1330#issuecomment-1962896565
